### PR TITLE
Experimental: Coverage without georef

### PR DESCRIPTION
**Pull request by azime**
Jira:
http://jira.canaltp.fr/browse/ITI-193

Will need merge of https://github.com/CanalTP/artemis/pull/244 right after

### CODE REVIEW

File: source/jormungandr/tests/journey_common_tests.py [hunk 1]
```diff
@@ -675,6 +676,23 @@ def test_novalidjourney_on_first_call(self):
         assert (feed_publisher["license"] == "L-contributor")
         assert (feed_publisher["url"] == "www.canaltp.fr")
 
+    def test_sp_outside_georf(self):
```
Comment (antoine-de) [inline source/jormungandr/tests/journey_common_tests.py]: typo on georef.

can you add a comment ?

File: source/georef/street_network.cpp [hunk 2]
```diff
@@ -403,11 +410,17 @@ PathFinder::find_nearest_stop_points(const navitia::time_duration& radius,
         dest_sp_idx.push_back(routing::SpIdx{e.first});
     }
     ProjectionGetterByCache projection_getter{mode, geo_ref.projected_stop_points};
-    return start_dijkstra_and_fill_duration_map<routing::SpIdx, routing::SpIdx, ProjectionGetterByCache>(
+    auto resp = start_dijkstra_and_fill_duration_map<routing::SpIdx, routing::SpIdx, ProjectionGetterByCache>(
             radius, dest_sp_idx, projection_getter);
+    for (auto r : resp) {
```
Comment (antoine-de) [inline source/georef/street_network.cpp]: const &

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 3]
```diff
@@ -75,13 +75,51 @@ def get_max_fallback_duration(request, mode):
     raise ValueError('unknown mode: {}'.format(mode))
 
 
+class Fallback(dict):
+    '''
+    Exmaple :
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: typo

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 4]
```diff
@@ -75,13 +75,51 @@ def get_max_fallback_duration(request, mode):
     raise ValueError('unknown mode: {}'.format(mode))
 
 
+class Fallback(dict):
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: I don't have better but I find this name a bit too generic :thinking: 

And the comment is a bit cryptic :stuck_out_tongue_winking_eye:

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 5]
```diff
@@ -75,13 +75,51 @@ def get_max_fallback_duration(request, mode):
...
+    def update_if_exist(self, mode, uri):
+        value = self.get(mode, None)
+        if value and uri in value:
+            self.update({uri: {"duration": 0, "status": response_pb2.reached}})
+
+    def get_duration(self, mode, uri):
+        values = self.get(mode)
+        return values.get(uri).get('duration')
+
+    def to_crowflay(self, mode, uri):
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: typo

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 6]
```diff
@@ -186,14 +224,21 @@ def _get_places_crowfly(instance, mode, place, max_duration_to_pt, max_nb_crowfl
 
     return {mode: res}
 
+def _get_duration(resp, place, mode, **kwargs):
+    map_response = {
+        response_pb2.unreached: -1,
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: do we want `-1` ? I think None is better no ?

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 7]
```diff
@@ -186,14 +224,21 @@ def _get_places_crowfly(instance, mode, place, max_duration_to_pt, max_nb_crowfl
...
+        response_pb2.reached: resp.duration,
+        response_pb2.unknown: int(place.distance/kwargs.get(mode))
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: we should add a comment here

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 8]
```diff
@@ -75,13 +75,51 @@ def get_max_fallback_duration(request, mode):
...
+
+    def get_duration(self, mode, uri):
+        values = self.get(mode)
+        return values.get(uri).get('duration')
+
+    def to_crowflay(self, mode, uri):
+        values = self.get(mode)
+        return values.get(uri).get("status") == response_pb2.unknown
+
+    def get_location_context(self, mode):
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: `return { k, v.get("duration") for k, v in self.get(mode, {}).items()`

File: source/kraken/worker.cpp [hunk 9]
```diff
@@ -919,11 +919,16 @@ void Worker::street_network_routing_matrix(const pbnavitia::StreetNetworkRouting
 
         auto* row = this->pb_creator.mutable_sn_routing_matrix()->add_rows();
         for(auto coord : dest_coords) {
+            auto* k = row->add_routing_response();
             auto it = nearest.find(coord.uri());
-            if (it != nearest.end()) {
-                row->add_duration(it->second.total_seconds());
-            }else {
-                row->add_duration(-1);
+            assert(it != nearest.end());
```
Comment (antoine-de) [inline source/kraken/worker.cpp]: why can you assert this ?

File: source/georef/street_network.cpp (hunk 2 above)
Comment (azime) [reply source/georef/street_network.cpp]: OK

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 7 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: OK

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 6 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: OK

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 5 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: ok

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 3 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: OK

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 8 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: ok

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 4 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: FallbackDuration

File: source/georef/street_network.cpp [hunk 10]
```diff
@@ -403,11 +410,17 @@ PathFinder::find_nearest_stop_points(const navitia::time_duration& radius,
         dest_sp_idx.push_back(routing::SpIdx{e.first});
     }
     ProjectionGetterByCache projection_getter{mode, geo_ref.projected_stop_points};
-    return start_dijkstra_and_fill_duration_map<routing::SpIdx, routing::SpIdx, ProjectionGetterByCache>(
+    auto resp = start_dijkstra_and_fill_duration_map<routing::SpIdx, routing::SpIdx, ProjectionGetterByCache>(
```
Comment (pbougue) [inline source/georef/street_network.cpp]: line length

File: source/georef/street_network.cpp [hunk 11]
```diff
@@ -403,11 +410,17 @@ PathFinder::find_nearest_stop_points(const navitia::time_duration& radius,
...
             radius, dest_sp_idx, projection_getter);
+    for (const auto& r : resp) {
+        if (r.second.routing_status == RoutingStatus_e::reached) {
```
Comment (pbougue) [inline source/georef/street_network.cpp:420]: @TeXitoi : maybe we should also send the `RoutingStatus_e::unknown` here? (maybe even all)
So for example on SNCF if all are unknown, jormun choose to send crowfly for all (at least jormun can choose).
Any opinion?

File: source/kraken/worker.cpp (hunk 9 above)
Comment (pbougue) [reply source/kraken/worker.cpp]: maybe throw a recoverable exception so we can see it in logs in release

**Review by pbougue — COMMENTED** [review]
Kraken read, I'd like to read tests and jormun after

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 12]
```diff
@@ -75,13 +75,46 @@ def get_max_fallback_duration(request, mode):
...
 
+class FallbackDuration(dict):
+    """
+    Dictionary containing the id of the places and the duration to arrive there and the projection status
+    Example :
+    {
+        "walking" : {"id1" : {"duration": 6, "status": response_pb2.reached},
+                    "id2" : {"duration": 15, "status": response_pb2.unknown}}}
+    """
+    def update_if_exist(self, mode, uri):
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: it's more a `init_if_exist` no? (or `reset_if_exist`)

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 13]
```diff
@@ -75,13 +75,46 @@ def get_max_fallback_duration(request, mode):
...
+
+    def get_duration(self, mode, uri):
+        values = self.get(mode)
+        return values.get(uri).get('duration')
+
+    def to_crowfly(self, mode, uri):
+        values = self.get(mode)
+        return values.get(uri).get("status") == response_pb2.unknown
+
+    def get_location_context(self, mode):
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: I'd rename it to get_location`s`_context`s`

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 14]
```diff
@@ -75,13 +75,46 @@ def get_max_fallback_duration(request, mode):
...
+    def update_if_exist(self, mode, uri):
+        value = self.get(mode, None)
+        if value and uri in value:
+            self.update({uri: {"duration": 0, "status": response_pb2.reached}})
+
+    def get_duration(self, mode, uri):
+        values = self.get(mode)
+        return values.get(uri).get('duration')
+
+    def to_crowfly(self, mode, uri):
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: maybe more `is_crowfly_needed`

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 15]
```diff
@@ -75,13 +75,46 @@ def get_max_fallback_duration(request, mode):
...
+        values = self.get(mode)
+        return values.get(uri).get("status") == response_pb2.unknown
+
+    def get_location_context(self, mode):
+        return {key: value.get("duration") for key, value in self.get(mode, {}).items()}
+
+    def merge(self, mode, values):
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: maybe rename it to `merge_reached_values`

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 16]
```diff
@@ -187,13 +220,22 @@ def _get_places_crowfly(instance, mode, place, max_duration_to_pt, max_nb_crowfl
     return {mode: res}
 
 
+def _get_duration(resp, place, mode, **kwargs):
+    map_response = {
+        response_pb2.reached: resp.duration,
+        # Calculate duration
+        response_pb2.unknown: int(place.distance/kwargs.get(mode))
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: no Manhattan on that?
we do one in new_default: https://github.com/CanalTP/navitia/blob/d91d4264681c3cd7b6d3edf4515b199ead5f8275/source/georef/street_network.cpp#L386

**Review by pbougue — COMMENTED** [review]
no tests on "unreached" case but not sure it's easy with navitia fixtures

File: source/georef/street_network.cpp (hunk 11 above)
Comment (TeXitoi) [reply source/georef/street_network.cpp:420]: I suspect this function is used directly in raptor_api, and that crow fly handling is managed another way. But I don't see the point about jormungandr.

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 17]
```diff
@@ -135,7 +171,7 @@ def _update_crowfly_duration(instance, mode, requested_entry_point):
         for stop_point in odt_sps:
             fallback_list[mode][stop_point.uri] = 0
             odt_stop_points.add(stop_point.uri)
-
+
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: trailing spaces

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 18]
```diff
@@ -388,32 +433,34 @@ def _build_from(self, journey, first_section, _from, crowfly_stop_points, odt_st
...
-        journey.departure_date_time = journey.departure_date_time - origins_fallback[departure.uri]
+        journey.departure_date_time = journey.departure_date_time - \
+                                      origins_fallback.get_duration(dep_mode, departure.uri)
         if _from.uri != departure.uri:
             if departure.uri in odt_stop_points:
                 journey.sections[0].origin.CopyFrom(_from)
-            elif departure.uri in crowfly_stop_points:
-                journey.sections.extend([create_crowfly(_from, departure, journey.departure_date_time,
-                                                        journey.sections[0].begin_date_time)])
+            elif departure.uri in crowfly_stop_points or origins_fallback.to_crowfly(dep_mode, departure.uri):
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: can't crowfly_stop_points and origins_fallback.to_crowfly be unified?

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 19]
```diff
@@ -434,14 +481,14 @@ def get_fallback_direct_path_futures(self, map_response, crowfly_stop_points, od
...
+            departure_datetime = journey.departure_date_time - g.origins_fallback.get_duration(dep_mode, departure.uri)
             # In the following cases, we don't need to compute the fallback direct path:
             # 1. the origin of the first section and the requested_origin are the same
             # 2. the origin of the first section belongs to a stop_area
             # 3. the origin of the first section belongs to odt stop_points
             if g.requested_origin.uri != departure.uri and \
-                            departure.uri not in odt_stop_points and \
-                            departure.uri not in crowfly_stop_points:
+                departure.uri not in odt_stop_points and \
+                    departure.uri not in crowfly_stop_points:
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: Strange indentation

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 20]
```diff
@@ -455,10 +502,10 @@ def get_fallback_direct_path_futures(self, map_response, crowfly_stop_points, od
             reverse_sections = False
             # In some cases, we don't need to compute the fallback direct path
             # Similar reasoning as above
-            arrival_datetime = journey.arrival_date_time + g.destinations_fallback[arr_mode][arrival.uri]
+            arrival_datetime = journey.arrival_date_time + g.destinations_fallback.get_duration(arr_mode, arrival.uri)
             if g.requested_destination.uri != arrival.uri and \
-                            arrival.uri not in odt_stop_points and \
-                            arrival.uri not in crowfly_stop_points:
+                arrival.uri not in odt_stop_points and \
+                    arrival.uri not in crowfly_stop_points:
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: Strange indentation

File: source/jormungandr/tests/journey_common_tests.py (hunk 1 above)
Comment (azime) [reply source/jormungandr/tests/journey_common_tests.py]: ok

File: source/georef/street_network.cpp (hunk 10 above)
Comment (azime) [reply source/georef/street_network.cpp]: OK

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 12 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: OK

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 14 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: OK

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 15 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: OK

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 17 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: OK

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 19 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: OK

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 20 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: OK

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 13 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: OK

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 21]
```diff
@@ -75,13 +78,46 @@ def get_max_fallback_duration(request, mode):
...
+
+    def is_crowfly_needed(self, mode, uri):
+        values = self.get(mode)
+        return values.get(uri).get("status") == response_pb2.unknown
+
+    def get_locations_contexts(self, mode):
+        return {key: value.get("duration") for key, value in self.get(mode, {}).items()}
+
+    def merge_reached_values(self, mode, values):
+        local_values = self.get(mode, None)
```
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: `None` is not necessary

Comment (azime) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: OK

**Review by antoine-de — APPROVED** [review]

**Comment by azime** [comment]
Artemis valid on this PR

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 16 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: OK

File: source/kraken/worker.cpp (hunk 9 above)
Comment (azime) [reply source/kraken/worker.cpp]: OK

File: source/jormungandr/tests/routing_tests_experimental.py [hunk 22]
```diff
@@ -109,6 +109,45 @@ def test_other_filtering(self):
...
+            "bike": instance.bike_speed,
+            "car": instance.car_speed,
+            "bss": instance.bss_speed,
+        }
+        request = {
+            "walking_speed": instance.walking_speed,
+            "bike_speed": instance.bike_speed,
+            "car_speed": instance.car_speed,
+            "bss_speed": instance.bss_speed,
+
```
Comment (pbougue) [inline source/jormungandr/tests/routing_tests_experimental.py]: very minor: empty line

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 23]
```diff
@@ -388,32 +443,34 @@ def _build_from(self, journey, first_section, _from, crowfly_stop_points, odt_st
...
-                                                        journey.sections[0].begin_date_time)])
+            elif departure.uri in crowfly_stop_points or origins_fallback.is_crowfly_needed(dep_mode, departure.uri):
+                journey.sections.extend([create_crowfly(journey, _from, departure, journey.departure_date_time,
+                                         journey.sections[0].begin_date_time)])
             else:
                 # extend the journey with the fallback routing path
                 self._extend_journey(journey, dep_mode, _from, departure, journey.departure_date_time,
                                      origins_fallback, True, fallback_direct_path)
-        journey.sections.sort(SectionSorter())
+        journey.sections.sort(SectionSorter())
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: minor: extra space

**Pull request merged**